### PR TITLE
issue #332. Fix algorithm argument parsing

### PR DESCRIPTION
**Pull request by rupertford**
The invoke-call kernel-argument parsing was only working for basic literals e.g. 1.0, 1_i_def and was rejecting the case where the literal value was negated e.g. -1.0, which was causing PSyclone to fail on some LFRic algorithm code. This PR fixes this problem and also adds support for more complex options e.g. 1.0*1.0, (1.0*1.0)/1.0.

### CODE REVIEW

File: src/psyclone/parse/algorithm.py [hunk 1]
```diff
@@ -594,9 +598,28 @@ def get_kernel(parse_tree, alg_filename):
...
             full_text = argument.tostr().lower()
             var_name = create_var_name(argument).lower()
             arguments.append(Arg('variable', full_text, var_name))
+        elif isinstance(argument, (Level_2_Unary_Expr, Add_Operand,
+                                   Parenthesis)):
+            # An expression e.g. -1, 1*n, ((1*n)/m). Note, for some
+            # reason Add_Operation represents binary expressions in
+            # fparser2.  Walk the tree to look for an argument.
+            found_arg = False
+            for statement in walk_ast([argument]):
```
Comment (arporter) [inline src/psyclone/parse/algorithm.py]: You could do `walk_ast([argument], [Name])` to just get all nodes that are instances of Name.

File: src/psyclone/parse/algorithm.py [hunk 2]
```diff
@@ -594,9 +598,28 @@ def get_kernel(parse_tree, alg_filename):
...
+                if isinstance(statement, Name):
+                    found_arg = True
+            if not found_arg:
+                arguments.append(Arg('literal', argument.tostr().lower()))
```
Comment (arporter) [inline src/psyclone/parse/algorithm.py]: Please comment, e.g.: "Store the string representation of the expression" (if that's what you're doing).

File: src/psyclone/parse/algorithm.py [hunk 3]
```diff
@@ -594,9 +598,28 @@ def get_kernel(parse_tree, alg_filename):
...
+            else:
+                raise NotImplementedError(
+                    "algorithm.py:get_kernel: Expressions containing "
+                    "arguments are not yet supported '{0}', value '{1}', "
```
Comment (arporter) [inline src/psyclone/parse/algorithm.py]: "containing arguments" should be "containing variables"?

File: src/psyclone/tests/parse/algorithm_test.py [hunk 4]
```diff
@@ -179,6 +179,76 @@ def test_getkernel_invalid_arg(parser, monkeypatch):
...
+    assert arg.text == content
+    assert arg.varname == "a_rg"
+
+
+@pytest.mark.parametrize('content',
+                         ["- arg", "1.0 * arg", "(1.0 * arg)",
+                          "1.0 * (1.0 * arg)", "arg1*arg2"])
+def test_getkernel_noexpr(parser, content):
+    '''Test that the get_kernel function recognises an expression
+    containing an argument and raises an exception (as this is not
```
Comment (arporter) [inline src/psyclone/tests/parse/algorithm_test.py]: Again, s/argument/variable/ I would have thought?

File: src/psyclone/parse/algorithm.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/parse/algorithm.py]: Great. I've simplified.

File: src/psyclone/parse/algorithm.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/parse/algorithm.py]: It is :-)

File: src/psyclone/parse/algorithm.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/parse/algorithm.py]: Oops.

File: src/psyclone/tests/parse/algorithm_test.py (hunk 4 above)
Comment (rupertford) [reply src/psyclone/tests/parse/algorithm_test.py]: At least I was consistent, even if I was wrong.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Pull request merged**